src/PostgREST/Workers.hs

```diff
@@ -99,18 +99,13 @@ connectionWorker appState = do
           AppState.putIsWorkerOn appState False
 
 -- | Check if a connection from the pool allows access to the PostgreSQL
--- database.  If not, the pool connections are released and a new connection is
--- tried.  Releasing the pool is key for rapid recovery. Otherwise, the pool
--- timeout would have to be reached for new healthy connections to be acquired.
--- Which might not happen if the server is busy with requests. No idle
--- connection, no pool timeout.
+-- database.
 --
 -- The connection tries are capped, but if the connection times out no error is
 -- thrown, just 'False' is returned.
 connectionStatus :: AppState -> IO ConnectionStatus
 connectionStatus appState =
-  retrying retrySettings shouldRetry $
-    const $ SQL.release pool >> getConnectionStatus
```

Comment:
Hmm.. one thing that was useful with releasing the pool. pg role settings kick in at login time, so by doing `ALTER ROLE rol SET statement_timeout = 5` and then `SIGUSR1`(or `NOTIFY pgrst`) we enforced the new `statement_timeout` for all pool connections.

I think it would be possible to apply the role settings ourselves but it'd be easier to let pg do that.

Why is necessary to remove `SQL.release`? Doing that cause errors now?

Reply:
Yes, `SQL.release` essentially destroys the pool.

Without patching `hasql-pool`, getting the same behaviour would require replacing the pool. That's quite possible but a bit of a hassle, I would have avoided it if not necessary.

E.g. something like
- in `AppState`, have a `TVar Pool` instead of `Pool`
- where we currently call `release`, instead
  1. create new pool
  2. place it in the tvar
  3. release the old pool

Do you think that's a good way to go? An alternative could be to patch `hasql-pool` to provide a variant of `release` (`flush`?) that keeps the pool alive and merely closes idle connections.

(It would be good to have a dedicated test for the `statement_timeout` functionality if that's a feature we want -- might be this test is basically that, but by its name etc. it seems to be aiming at a different feature. I'm imagining a test similar to this one that
- runs something like `pg_sleep(n)` successfully
- lowers the statement timeout to `< n` similarly to this test, then sends `USR1`
- and now the same `pg_sleep(n)` rpc times out
Does that seem sensible?)

Reply:
> 1. Without patching hasql-pool, getting the same behaviour would require replacing the pool. That's quite possible but a bit of a hassle, I would have avoided it if not necessary.
> 2. An alternative could be to patch hasql-pool to provide a variant of release (flush?) that keeps the pool alive and merely closes idle connections.

Yeah 1 seems too complicated. 2 wouldn't help because we'd need to close all connections(not only idle ones) to keep the same behavior.

> I think it would be possible to apply the role settings ourselves but it'd be easier to let pg do that.

Another option would be getting the statement_timeout from `pg_db_role_settings` and applying it at the session level ourselves. To reset the current `statement_timeout`, perhaps we can run a `SET statement_timeout` on all connections when `SIGUSR1/NOTIFY` is done.

Reply:
> runs something like pg_sleep(n) successfully
> lowers the statement timeout to < n similarly to this test, then sends USR1
> and now the same pg_sleep(n) rpc times out
> Does that seem sensible?)

Yeah, that sounds good.

> Another option would be getting the statement_timeout from pg_db_role_settings and applying it at the session level ourselves.

I can help with that one btw, if there's no simpler alternative.

Reply:
> 2 wouldn't help because we'd need to close all connections(not only idle ones) to keep the same behavior.

Just to clarify this one point: Currently, does releasing the pool interrupt any in-progress connections to the database? At first glance that would seem like a bit of a questionable design to me... E.g. that would mean HTTP error responses for requests that might have resulted in a successful database update. (Which isn't terrible, just something I'd have thought we'd avoid for this kind of configuration reload.) Intuitively, I wouldn't expect a statement timeout configuration change to apply to in-progress requests, just to requests that start after the change.

It does make sense to me that we'd want the configuration change to take effect immediately, so yeah the current state of the PR is no good, we need to close at least the idle connections on `SIGUSR1`.

Reply:
>  Currently, does releasing the pool interrupt any in-progress connections to the database?

Yes, this can be reproduced with a sleep function and a pool size of 1, doing a SIGUSR1 will then abort the sleep request.

> I wouldn't expect a statement timeout configuration change to apply to in-progress requests, just to requests that start after the change.

Hm, I think this would be a better behavior as well. Though in a busy server with all connections taken, a `flush` function wouldn't find any idle connections and then no new `statement_timeout` would be applied right? Or there could be a way to queue this `flush` operation?

Reply:
> Yes, this can be reproduced with a sleep function and a pool size of 1, doing a SIGUSR1 will then abort the sleep request.

Wait, I'm wrong. The above is not true if doing a `notify pgrst, 'reload schema';`(which also calls `release`) on a pool size of 2 and a request with `sleep`. Something wrong with my test, SIGUSR1 is somehow killing the database.

Reply:
So current `release` is indeed not interrupting in-progress connections to the db. I was wrong all along :man_facepalming: .

Reply:
The updated version addresses this -- we now have a dedicated `flushPool` called here which replaces the pool with a fresh one.

Reply:
@robx Really sorry for the misdirection, I forgot about a majorly important detail here.

It's critical that [Automatic Schema Cache Reloading](https://postgrest.org/en/latest/schema_cache.html#automatic-schema-cache-reloading) can be done continuously on a production instance. For that, we can't have ​a `NOTIFY pgrst, 'reload schema'` or `SIGUSR1` interrupting running requests(commented that pending test in https://github.com/PostgREST/postgrest/issues/1766).
 
Under that restriction, it'd be preferable to lose the eventual application(due to the old pool timeout) of the new `statement_timeout`. 

Another option I see would be to move the `flushPool` to another unix signal like SIGHUP(we had trouble with using this one before). However, I don't see the real benefit over a `systemctl restart postgrest` since we'll be inoperative anyway.

It might be possible to apply `statement_timeout` with session level settings later. So it's not that bad to lose this now.

Reply:
This doesn't interrupt running requests, though. What `flushPool` does is 1. create a new pool 2. switch out the pools in the `IORef`, so any new requests will be handled by the new pool 3. releases the old pool, which just means that idle connections from the old pool are closed.

Running requests aren't affected; the `use` call for a running request will just find that its pool has been released, so the connection is closed instead of returning it to the pool.

Reply:
That is awesome. Can that be remarked as a comment on the `flushPool` function? A test that confirms we don't interrupt running requests would also be really nice.

Reply:
Comment for posterity, I was being fooled by this manual test:

```
PGRST_DB_ANON_ROLE='postgrest_test_anonymous' postgrest-with-postgresql-14
postgrest-run

05/Aug/2022:12:07:49 -0500: Attempting to connect to the database...
05/Aug/2022:12:07:49 -0500: Connection successful
05/Aug/2022:12:07:49 -0500: Listening on port 3000
05/Aug/2022:12:07:49 -0500: Config reloaded
05/Aug/2022:12:07:49 -0500: Listening for notifications on the pgrst channel
05/Aug/2022:12:07:49 -0500: Schema cache loaded

curl localhost:3000/rpc/sleep

# on another terminal I did
# kill -USR1 $(pgrep postgrest)
# and then the request was interrupted

{"code":"PGRST001","details":"no connection to the server\n","hint":null,"message":"Database client error. Retrying the connection."}
```

My mistake was omitting the `-n(newest)` flag in pgrep(should have been a `pgrep -n postgrest`) and I was killing postgresql instead. TIL that [SIGUSR1 triggers a quickdie for pg](https://www.postgresql.org/docs/7.0/signals.htm).

Reply:
> 2. switch out the pools in the IORef, so any new requests will be handled by the new pool
> 3. releases the old pool, which just means that idle connections from the old pool are closed.

Question here, does that mean the old pool connections will continue to live on when the running requests are finished? Now that we don't have the idle timeout those will not be closed right?

Reply:
They'll be closed as soon as their requests are done. When `use` tries to return them to the released pool, it closes them and discards them instead.